src/Chapter2.hs

```diff
@@ -728,7 +746,14 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate [] = []
+smartReplicate (x:xs) = replicate x x ++ smartReplicate xs
+
+{- How can we do this with map + replicate? The closes I got was:
+smartReplicate xs = concat (map (replicate x) xs)
+where x = head xs
+, but that just replicates each element by the first elements value.
+Is there a way to get the current element value when using map? -}
```

Comment:
I hope the previous comment answers this question, but let me know if you'd like to discuss it more detailed 🙂

Reply:
Would really like to get to the bottom of it - I understand your comment above, however, what I can't figure out is what value to pass to the partial `replicate` function that will be used to map over the list:

`smartReplicate xs = concat (map (replicate x) xs)`

I think this is along the right lines, but how do I specify what the correct `x` is?

Reply:
You need to pass to the `replicate` function the current element you are on while mapping the list, so it may look like this:

```haskell
smartReplicate xs = concat (map (\x -> replicate x x) xs)
```

This mean, that we only need to pass the function `replicate to `map`. Remember that the `map` has the following type:

```haskell
map :: (a -> b) -> [a] -> [b]
```

So we need to give it just a function that somehow transforms the current single element in the list. It has the type `a -> b` and in our case , we could think of such function as the following `f` function:

```haskell
f :: Int -> [Int]
f I = replicate i i
```

Remember that `replicate :: Int -> Int -> [Int]` so the types are correct.
And as I mentioned above, you can use `concatMap` instead of `concat` and `map`:

```haskell
smartReplicate = concatMap f
  where
    f :: Int -> [Int]
    f I = replicate i i
```

Does this makes sense?